### PR TITLE
WIP: Check for sufficient channel buffer size for signals being trapped

### LINKED ISSUE #30: Check if channel provided to signal.Notify has sufficient buffer
https://golang.org/pkg/os/signal/#Notify defines:

```
Package signal will not block sending to c: the caller must ensure that c has sufficient buffer
space to keep up with the expected signal rate. For a channel used for notification of just one
signal value, a buffer of size 1 is sufficient.
```

Should staticcheck check any channel provided to os.Notify has a buffer of at least the number of signals being trapped?

Note this makes it slightly more complicated to reduce false positives, to do it effectively (and without thinking too much about it), do you have any immediate thoughts on the best way to track all calls to notify for a channel and then reducing them for each stop?

```
It is allowed to call Notify multiple times with the same channel: each call expands the set of
signals sent to that channel. The only way to remove signals from the set is to call Stop.
```

Idea by @judwhite

### PR DESCRIPTION
**Pull request by bradleyfalzon**
I'm concerned about the actual implementation itself, I'm not certain whether this is the most optimal.

The check currently ensures the buffer size == number of signals caught, however, in discussions with @judwhite and re-reading docs, I also believe just a check the channel is buffered is enough. As the buffer size above 1 depends on the signal rate, not the number of signals.

A test still fails until we can agree whether the check should only check for a buffered channel, or whether it should check the buffer size is at least number of signals being trapped.

Fixes #30

### CODE REVIEW

File: lint.go [hunk 1]
```diff
@@ -91,6 +93,98 @@ func hasType(f *lint.File, expr ast.Expr, name string) bool {
...
+			}
+			obj := f.Pkg.TypesInfo.ObjectOf(chn)
+			if obj == nil {
+				return true
+			}
+			if buffSize, ok := chanPosSize[obj.Pos()]; ok {
+				if buffSize < len(call.Args)-1 {
+					f.Errorf(chn, "channel buffer size %d is too small to catch %v signal(s)", buffSize, len(call.Args)-1)
+				}
+			}
```
Comment (judwhite) [inline lint.go:149]: Noting where the change to `>= 1` would be if that's the route to take.

**Comment by bradleyfalzon** [comment]
If we're only checking channels are buffered, then we could also just check calls to make with an os.Signal have a buffer regardless where they're used.

This could open it up for false positives though, if someone is creating the channels for some other purpose, but simplifies the test, improves speed and reduces false negatives when a unbuffered channel is passed into a function calling signal.Notify (this check doesn't handle that).

**Comment by dominikh** [comment]
Please don't simplify the logic. In fact, I would recommend putting the PR on hold for now. By the time I am done with implementing VRP, checking a channel's length will be trivial.

**Comment by bradleyfalzon** [comment]
Sounds good to me.

**Comment by bradleyfalzon** [comment]
Closing for the moment as an alt implementation will be better.

**Pull request closed without merging**